---
title: "Assemblycomparator2 Report"
subtitle: "Developed by Oliver Hansen & Carl M. Kobel"
date: "`r Sys.Date()`"
output:
  prettydoc::html_pretty:
    theme: cayman
    highlight: github
editor_options: 
  chunk_output_type: console
---
  
  
```{r echo=FALSE, message=F, warning=F}
#Dependencies
library(tidyverse)
library(prettydoc)
#library(broom)
#library(magrittr)
#library(knitr)
#library(kableExtra)
library(DT)

# This is useful for debugging the markdown script.
debug_dir ="~/assemblycomparator2/tests/E._faecium/output_asscom2/"
setwd(debug_dir)
```


```{r echo=FALSE, message=F, warning=F}



# Import the metadata table
# This table makes it easy for us later to convert the long paths to short and simple sample names
metadata_file = "metadata.tsv"

if (file.exists(metadata_file)) {
    metadata_df = read_tsv(metadata_file)
    
    metadata_df %>%
        #arrange(sample) %>% 
        datatable(class = 'cell-border stripe') %>%
        print()
    
    
    # Generate a very simple table which can be used to convert long paths to sample names
    metadata_translate = metadata_df %>% 
        select(sample, filename = input_file_fasta)
    
} else {
    stop(paste("The metadata file", metadata_file, "is not accessible."))
}


```

#assembly-stats overview

Assembly statistics provided by https://github.com/sanger-pathogens/assembly-stats

```{r echo=FALSE, message=F, warning=F}
assembly_stats_file = "assembly-stats/assembly-stats.tsv"
if (file.exists(assembly_stats_file)) {
    assembly_stats_df = read_tsv(assembly_stats_file) 
    
    assembly_stats_df %>% left_join(metadata_translate) %>% 
        select(sample, everything(), -filename) %>% 
        arrange(sample) %>% 
        datatable(class = 'cell-border stripe')
            
    
    
} else {
    paste("Warning: The file", assembly_stats_file, "is not accessible.")
}
```



```{r fig.height=100, fig.width=100}
#MLST overview

#The mlst software incorporates components of the PubMLST database which must be cited in any publications that use mlst:

#"This publication made use of the PubMLST website (https://pubmlst.org/) developed by Keith Jolley (Jolley & Maiden 2010, BMC Bioinformatics, 11:595) and sited at the University of Oxford. The development of that website was funded by the Wellcome Trust".

#You should also cite this software (currently unpublished) as:
#Seemann T, mlst Github https://github.com/tseemann/mlst
```
## MLST
```{r echo=FALSE, message=F, warning=F}
#f√• mlst til at spytte header ud.
#mlst <- read.table(file = 'mlst.tsv', sep = '\t', header = FALSE)
mlst_file = "mlst/mlst.tsv"
if (file.exists(mlst_file)) {
    mlst_df <- read_tsv("mlst/mlst.tsv", col_names = F)
    allelnum <- length(mlst_df) - 3
    names(mlst_df) = c("filename", "scheme", "sequence_type", paste('allele_', 1:allelnum))
    
    
    mlst_df %>%
        left_join(metadata_translate) %>% 
        select(sample, everything(), -filename) %>% 
        arrange(sample) %>% 
        datatable(class = 'cell-border stripe')
        
    
    #mlst_df %>% datatable(
    #  options = list(
    #   autoWidth = TRUE, 
    #   columnDefs = list(list(width = '300px'))),
      #class = 'cell-border stripe',
      #rownames = FALSE,
      #colnames = c('Isolate','Scheme', 'SequenceType', rep('Allel',allelnum) )
      #)
    
      #kable() %>% 
      #kable_paper('hover', full_width = F)
}
```

```{r echo=FALSE, message=F, warning=F}
##Abricate
#It only supports contigs, not FASTQ reads
#It only detects acquired resistance genes, NOT point mutations
#It uses a DNA sequence database, not protein
#It needs BLAST+ >= 2.7

#uses the 3 databases:
#plasmidfinder
#Ncbi
#card

```


## NCBI Resistance genes

```{r echo=FALSE, message=F, warning=F}

abricate_ncbi_file = "abricate/ncbi_summarized.tsv"
if (file.exists(abricate_ncbi_file)) {
    abricate_ncbi_df = read_tsv(abricate_ncbi_file) %>% 
        rename(filename = `#FILE`, num_found = NUM_FOUND)
    
    
    abricate_ncbi_df %>%
        left_join(metadata_translate) %>% 
        select(sample, everything(), -filename) %>% 
        arrange(sample) %>% 
        datatable(class = 'cell-border stripe')

    
} else {
    paste("Warning: The file", abricate_ncbi_file, "is not accessible.")
}
```

## Card Resistance genes

```{r echo=FALSE, message=F, warning=F}

abricate_card_file = "abricate/card_summarized.tsv"
if (file.exists(abricate_card_file)) {
    abricate_card_df = read_tsv(abricate_card_file) %>% 
        rename(filename = `#FILE`, num_found = NUM_FOUND)
    
    
    abricate_card_df %>%
        left_join(metadata_translate) %>% 
        select(sample, everything(), -filename) %>% 
        arrange(sample) %>% 
        datatable(class = 'cell-border stripe')

    
} else {
    paste("Warning: The file", abricate_card_file, "is not accessible.")
}
```

## Plasmidfinder calls

```{r echo=FALSE, message=F, warning=F}

abricate_plasmidfinder_file = "abricate/plasmidfinder_summarized.tsv"
if (file.exists(abricate_plasmidfinder_file)) {
    abricate_plasmidfinder_df = read_tsv(abricate_plasmidfinder_file) %>% 
        rename(filename = `#FILE`, num_found = NUM_FOUND)
    
    
    abricate_plasmidfinder_df %>%
        left_join(metadata_translate) %>% 
        select(sample, everything(), -filename) %>% 
        arrange(sample) %>% 
        datatable(class = 'cell-border stripe')

    
} else {
    paste("Warning: The file", abricate_plasmidfinder_file, "is not accessible.")
}
```

```{r}



# 
# plasmidfinder_detail <- read.table(file = 'abricate/plasmidfinder_detailed.tsv', sep = '\t', header = TRUE)
# 
# 
# plasmidfinder_sum <- read.table(file = 'abricate/plasmidfinder_summarized.tsv', sep = '\t', header = TRUE)
# 
# plasmidfinder_sum %>% datatable(
#   options = list(
#    autoWidth = TRUE, 
#    columnDefs = list(list(width = '50px'))),
#   class = 'cell-border stripe',
#   rownames = FALSE
#     )
# 
# plasmidfinder_detail  %>% datatable(
#   options = list(
#    autoWidth = TRUE, 
#    columnDefs = list(list(width = '50px'))),
#   class = 'cell-border stripe',
#   rownames = FALSE 
#  )
```

##Stats for the card database
