---
title: "Assemblycomparator2 Report"
#subtitle: "`r head(tail(unlist(strsplit(getwd(), "/")) ,2) ,1)`" # use the title variable from the pipeline instead somehow
date: "`r Sys.Date()`"
#output:
  #prettydoc::html_pretty:
    #theme: cayman
    #highlight: github
editor_options: 
  chunk_output_type: console
#css: "max-width: 5000px; margin: auto; padding: 1em; line-height: 20px"
---
  
  
```{r echo=FALSE, message=F, warning=F}
#Dependencies
library(tidyverse)
library(prettydoc)
#library(broom)
#library(magrittr)
#library(knitr)
#library(kableExtra)
library(DT)

# This is useful for debugging the markdown script.
#debug_dir ="~/assemblycomparator2/tests/E._faecium/output_asscom2/"
#setwd(debug_dir)

# A function that returns a new table number for each run
tableno_var = 0
tableno = function() {
    tableno_var <<- tableno_var + 1
    tableno_var
}

figno_var = 0
figno = function() {
    figno_var <<- figno_var + 1
    figno_var
}


```

## Sample overview
```{r echo=FALSE, message=F, warning=F}



# Import the metadata table
# This table makes it easy for us later to convert the long paths to short and simple sample names
metadata_file = "metadata.tsv"

if (file.exists(metadata_file)) {
    metadata_df = read_tsv(metadata_file)
    
    
    # Generate a very simple table which can be used to convert long paths to sample names
    metadata_translate = metadata_df %>% 
        select(sample, filename = input_file_fasta)
    
    # Present the metadata in the report
    metadata_df %>%
        select(-index) %>% 
        arrange(sample) %>% 
        datatable(class = 'cell-border stripe')
    
    
    
} else {
    stop(paste("The metadata file", metadata_file, "is not accessible."))
}


```

## Assembly statistics

Assembly statistics is provided by [assembly-stats](https://github.com/sanger-pathogens/assembly-stats#assembly-stats).

```{r echo=FALSE, message=F, warning=F}
assembly_stats_file = "assembly-stats/assembly-stats.tsv"
if (file.exists(assembly_stats_file)) {
    assembly_stats_df = read_tsv(assembly_stats_file) 
    
    assembly_stats_df %>% left_join(metadata_translate) %>% 
        select(sample, everything(), -filename) %>% 
        arrange(sample) %>% 
        datatable(class = 'cell-border stripe')
            
    
    
} else {
    paste("Warning: The file", assembly_stats_file, "is not accessible.")
}
```

Lengths of contigs:

```{r echo=FALSE, message=F, warning=F}
# This chunk is needed for setting the size of the contig lengths plot

sequence_lengths_file = "sequence_lengths/sequence_lengths.tsv"
if (file.exists(sequence_lengths_file)) {
    sequence_lengths_df = read_tsv(sequence_lengths_file, col_names = c("sample", "record", "length"))

    
    fig_height = max(2.28,
                     (sequence_lengths_df$sample %>% length) *0.12)
    
    knitr::opts_chunk$set(fig.height = fig_height,
                          fig.width = 10)
}
```

```{r echo=FALSE, message=F, warning=F}

# Now when the figure height has been set, we can generate the figure
if (file.exists(sequence_lengths_file)) {
    sequence_lengths_df %>%
        group_by(sample) %>% 
        arrange(length) %>% 
        
        mutate(col = row_number(-length)) %>% 
        mutate(init = lag(length, default = 0),
               start = cumsum(init),
               stop = start + length - 1) %>% 
        
        ggplot(aes(fill=factor(col), y=sample, x=length)) + 
    geom_bar(position="dodge", stat="identity") +
        
        #scale_x_log10() +
        theme_classic() +
        theme(legend.position = "none")
        
        #ggplot() + 
        #geom_segment(aes(x = start, xend = stop, y = sample, yend = sample, color = factor(col)), size = 2)
            
    
    
} else {
    paste("Warning: The file", sequence_lengths_file, "is not accessible.")
}
```



## Species

Species identification is provided by [Kraken 2](https://github.com/DerrickWood/kraken2/wiki/About-Kraken-2).




```{r echo=FALSE, message=F, warning=F}

kraken2_file = "kraken2/kraken2_reports.tsv"
if (file.exists(kraken2_file)) {
    
    
    
    
    kraken2_df = read_tsv(kraken2_file) %>% 
        rename(percent = match_percent)
    
    
    kraken2_df %>% group_by(sample) %>% 
        #filter(level == "S" | level == "U")
        filter(level == "S") %>% 
        mutate(rank = row_number(-percent),
               rank_text = paste("hit", rank),
               clade_text = paste0(percent, "% (", clade, ")")) %>% 
               #clade_text = paste0(clade, " ", match_percent, ")")) %>% 
        
        filter(rank <= 2) %>% 
        pivot_wider(id_cols = sample, names_from = rank_text, values_from = c(percent, clade)) %>% 
        
    
        arrange(sample) %>% 
    
        datatable(class = 'cell-border stripe')
            
    
    
} else {
    paste("Warning: The file", kraken2_file, "is not accessible.")
}
```
`r if (file.exists(kraken2_file)) {paste("Using the", Sys.getenv("ASSCOM2_KRAKEN2_DB"), "database")}`.





## MLST

Multi locus sequence typing provided by [mlst](https://github.com/tseemann/mlst#mlst).
The mlst software incorporates components of the [PubMLST](https://pubmlst.org/) database.
```{r echo=FALSE, message=F, warning=F}
#f√• mlst til at spytte header ud.
#mlst <- read.table(file = 'mlst.tsv', sep = '\t', header = FALSE)
mlst_file = "mlst/mlst.tsv"
if (file.exists(mlst_file)) {
    mlst_df <- read_tsv("mlst/mlst.tsv", col_names = F)
    allelnum <- length(mlst_df) - 3
    names(mlst_df) = c("filename", "scheme", "sequence_type", paste('al.', 1:allelnum))
    
    
    mlst_df %>%
        left_join(metadata_translate) %>% 
        select(sample, everything(), -filename) %>% 
        arrange(sample) %>% 
        datatable(class = 'cell-border stripe')
        
    
    #mlst_df %>% datatable(
    #  options = list(
    #   autoWidth = TRUE, 
    #   columnDefs = list(list(width = '300px'))),
      #class = 'cell-border stripe',
      #rownames = FALSE,
      #colnames = c('Isolate','Scheme', 'SequenceType', rep('Allel',allelnum) )
      #)
    
      #kable() %>% 
      #kable_paper('hover', full_width = F)
}
```

```{r echo=FALSE, message=F, warning=F}
##Abricate
#It only supports contigs, not FASTQ reads
#It only detects acquired resistance genes, NOT point mutations
#It uses a DNA sequence database, not protein
#It needs BLAST+ >= 2.7

#uses the 3 databases:
#plasmidfinder
#Ncbi
#card

```


## NCBI Resistance
Resistance genes provided by [NCBI AMRFinder](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6811410/) via [Abricate](https://github.com/tseemann/abricate#abricate).

```{r echo=FALSE, message=F, warning=F}

abricate_ncbi_file = "abricate/ncbi_summarized.tsv"
if (file.exists(abricate_ncbi_file)) {
    abricate_ncbi_df = read_tsv(abricate_ncbi_file) %>% 
        rename(filename = `#FILE`, num_found = NUM_FOUND)
    
    
    abricate_ncbi_df %>%
        left_join(metadata_translate) %>% 
        select(sample, everything(), -filename) %>% 
        arrange(sample) %>% 
        datatable(class = 'cell-border stripe')

    
} else {
    paste("Warning: The file", abricate_ncbi_file, "is not accessible.")
}
```

## Card Resistance genes

```{r echo=FALSE, message=F, warning=F}

abricate_card_file = "abricate/card_summarized.tsv"
if (file.exists(abricate_card_file)) {
    abricate_card_df = read_tsv(abricate_card_file) %>% 
        rename(filename = `#FILE`, num_found = NUM_FOUND)
    
    
    abricate_card_df %>%
        left_join(metadata_translate) %>% 
        select(sample, everything(), -filename) %>% 
        arrange(sample) %>% 
        datatable(class = 'cell-border stripe')

    
} else {
    paste("Warning: The file", abricate_card_file, "is not accessible.")
}
```

## Plasmidfinder calls

```{r echo=FALSE, message=F, warning=F}

abricate_plasmidfinder_file = "abricate/plasmidfinder_summarized.tsv"
if (file.exists(abricate_plasmidfinder_file)) {
    abricate_plasmidfinder_df = read_tsv(abricate_plasmidfinder_file) %>% 
        rename(filename = `#FILE`, num_found = NUM_FOUND)
    
    
    abricate_plasmidfinder_df %>%
        left_join(metadata_translate) %>% 
        select(sample, everything(), -filename) %>% 
        arrange(sample) %>% 
        datatable(class = 'cell-border stripe')

    
} else {
    paste("Warning: The file", abricate_plasmidfinder_file, "is not accessible.")
}
```
------

## Pan and Core genome

[Roary](https://sanger-pathogens.github.io/Roary/) the pan genome pipeline computes the number of orthologous genes in a number of core/pan partitions.

The core genome denotes the genes which are conserved between all samples (intersection), whereas the pan genome is the union of all genes across all samples.

```{r echo=FALSE, message=F, warning=F}
roary_summary_file = "roary/summary_statistics.txt"
if (file.exists(roary_summary_file)) {
    paste("yes")
    read_tsv(roary_summary_file, col_names = c("partition", "definition", "count")) %>% 
        mutate(type = paste(partition, definition)) %>% 
        select(type, count) %>% 
        datatable(class = "stripe", colnames = NULL, rownames = F,  options = list(dom = 't', ordering=F), width = 400)
}
```

*Table `r tableno()`: Distribution of genes in different core/pan partitions.*

------


```{r echo=FALSE, message=F, warning=F}
# This chunk is needed for setting the size of the roary gpa plot
fig_height = max(1.6,
                 (metadata_df$sample %>% length) *0.4)

knitr::opts_chunk$set(fig.height = fig_height,
                      fig.width = 8)
```


```{r echo = F, message = F, warning = F}
roary_gpa_file = "roary/gene_presence_absence.Rtab"
if (file.exists(roary_gpa_file)) {
    paste("yes")
    roary_gpa_df = read_tsv(roary_gpa_file)
    
    roary_gpa_df %>% 
        #select(-Gene) %>% 
        mutate(sum = rowSums(across(where(is.numeric)))) %>%
        #mutate(order = 1:(dim(.)[1])) %>% View
        mutate(genes = 1:length(Gene)) %>% 
        select(-Gene) %>% 
    
        pivot_longer(c(everything(), -genes, -sum), names_to = "sample", values_to = "present") %>% 
        filter(present >= 1) %>% 
        #View
        ggplot(aes(genes, sample, fill = factor(sum))) + geom_tile() +
        labs(fill = "share count") + 
        theme_classic() 
}
```

*Fig. `r figno()`: Visual distribution of genes in the different samples. The genes are ordered in respect to the number of samples sharing them.*



------

The [assemblycomparator2](https://github.com/cmkobel/assemblycomparator2#assemblycomparator2) pipeline and this report is developed by Oliver Hansen & Carl M. Kobel
